### PR TITLE
Bugfix: Websocket ping/pong improvements

### PR DESCRIPTION
**Pull request by gloriousCode**
Alt Text

# PR Description
This is a pretty basic PR which fixes up an issue where a ping message could theoretically crash the entire application due to a lack of locking. Turns out, only BTSE had the potential for this, but I've standardised ping handling a bit more while I was looking around. Checked each exchange websocket documentation to verify whether any ping pongs needed to be added or removed.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
- [x] go test ./... -race
- [x] golangci-lint run
- [x] `TestSetupPingHandler()`

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/kraken/kraken_websocket.go [hunk 1]
```diff
@@ -148,28 +151,16 @@ func (k *Kraken) WsHandleData() {
...
-			if err != nil {
-				k.Websocket.DataHandler <- err
-			}
-		}
+func (k *Kraken) wsPingHandler() error {
+	message, err := json.Marshal(pingRequest)
+	if err != nil {
+		return err
 	}
+	k.WebsocketConn.SetupPingHandler(wshandler.WebsocketPingHandler{
```
Comment (shazbert) [inline exchanges/kraken/kraken_websocket.go]: should set message type here

File: exchanges/websocket/wshandler/wshandler.go [hunk 2]
```diff
@@ -656,32 +655,80 @@ func (w *WebsocketConnection) Dial(dialer *websocket.Dialer, headers http.Header
...
+	return w.Connection.WriteMessage(messageType, message)
+}
+
+// SetupPingHandler will automatically send ping or pong messages based on
+// WebsocketPingHandler configuration
+func (w *WebsocketConnection) SetupPingHandler(handler WebsocketPingHandler) {
+	if handler.UseGorillaHandler {
+		h := func(msg string) error {
+			err := w.Connection.WriteControl(handler.MessageType, []byte(msg), time.Now().Add(handler.Delay))
+			if e, ok := err.(net.Error); ok && e.Temporary() {
```
Comment (shazbert) [inline exchanges/websocket/wshandler/wshandler.go]: should also test for elseif err == websocket.ErrCloseSent and return nil

File: exchanges/websocket/wshandler/wshandler_test.go [hunk 3]
```diff
@@ -602,6 +616,43 @@ func TestSendMessageWithResponse(t *testing.T) {
...
+	wc.Shutdown = make(chan struct{})
+	err := wc.Dial(&dialer, http.Header{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wc.SetupPingHandler(WebsocketPingHandler{
+		UseGorillaHandler: true,
+		MessageType:       websocket.TextMessage,
+		Message:           []byte(Ping),
```
Comment (shazbert) [inline exchanges/websocket/wshandler/wshandler_test.go]: rm not being used when using gorilla handler

File: exchanges/btse/btse_websocket.go [hunk 4]
```diff
@@ -33,7 +33,10 @@ func (b *BTSE) WsConnect() error {
 	if err != nil {
 		return err
 	}
-	go b.Pinger()
+	b.WebsocketConn.SetupPingHandler(wshandler.WebsocketPingHandler{
```
Comment (shazbert) [inline exchanges/btse/btse_websocket.go]: just making sure this doesn't need handler.Message

Comment (gloriousCode) [reply exchanges/btse/btse_websocket.go]: Not for this one
From the docs:
>  To ensure that your session remains active, send a Ping frame before you reach the 60 second timeout.

Previous implementation was:
`b.WebsocketConn.Connection.WriteMessage(websocket.PingMessage, nil)`
which is the same as what's happening here

File: exchanges/kraken/kraken_websocket.go (hunk 1 above)
Comment (thrasher-) [reply exchanges/kraken/kraken_websocket.go]: Just reporting that the latest commit fixes the [ERROR] | 03/01/2020 11:23:24 | Kraken failed to send message to websocket {"event":"ping"} error

File: exchanges/websocket/wshandler/wshandler.go [hunk 5]
```diff
@@ -656,32 +655,82 @@ func (w *WebsocketConnection) Dial(dialer *websocket.Dialer, headers http.Header
...
 	defer w.Unlock()
 	if !w.IsConnected() {
 		return fmt.Errorf("%v cannot send message to a disconnected websocket", w.ExchangeName)
 	}
-	json, err := json.Marshal(data)
-	if err != nil {
-		return err
+	if w.Verbose {
+		log.Debugf(log.WebsocketMgr,
+			"%v sending message to websocket %v", w.ExchangeName, data)
```
Comment (thrasher-) [inline exchanges/websocket/wshandler/wshandler.go]: Please use %+v for data

Comment (gloriousCode) [reply exchanges/websocket/wshandler/wshandler.go]: woah man, I never knew about that plus flag. 🎉 🎉 🎉 🎉 🎉

**Pull request merged**